streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -136,7 +145,29 @@ public void register(final StateStore store,
 
         log.info("Restoring state for global store {}", store.name());
         final List<TopicPartition> topicPartitions = topicPartitionsForStore(store);
-        final Map<TopicPartition, Long> highWatermarks = globalConsumer.endOffsets(topicPartitions);
+        Map<TopicPartition, Long> highWatermarks = null;
+
+        int attempts = 0;
+        while (highWatermarks == null) {
+            try {
+                highWatermarks = globalConsumer.endOffsets(topicPartitions);
+            } catch (final TimeoutException retryableException) {
+                if (++attempts > retries) {
+                    log.error("Failed to get end offsets for topic partitions of global store {} after " + retries + " retry attempts. " +
+                        "You can increase the number of retries via configuration parameter `retries`.",
+                        store.name(),
+                        retryableException);
+                    throw new StreamsException(String.format("Failed to get end offsets for topic partitions of global store %s after " + retries + " retry attempts. " +
```

Comment:
The string in the log and the exception are identical, extract into into a variable?

Reply:
It may be a bit tricky as they use different string concatenation methods.